src/types/redis_hyperloglog.cc

```diff
@@ -0,0 +1,401 @@
+ * pointed by 'invalid' is set to non-zero, otherwise it is left untouched.
+ *
+ * hllCount() supports a special internal-only encoding of HLL_RAW, that
+ * is, hdr->registers will point to an uint8_t array of HLL_REGISTERS element.
+ * This is useful in order to speedup PFCOUNT when called against multiple
+ * keys (no need to work with 6-bit integers encoding). */
+uint64_t Hyperloglog::hllCount(const std::vector<uint8_t> &counts) {
+  double m = HLL_REGISTERS;
+  double E;
+  int j;
+  /* Note that reghisto size could be just HLL_Q+2, because HLL_Q+1 is
+   * the maximum frequency of the "000...1" sequence the hash function is
+   * able to return. However it is slow to check for sanity of the
+   * input: instead we history array at a safe size: overflows will
+   * just write data to wrong, but correctly allocated, places. */
+  int reghisto[64] = {0};
+
+  /* Compute register histogram */
+  hllDenseRegHisto((uint8_t *)(&counts[0]), reghisto);
+
+  /* Estimate cardinality from register histogram. See:
+   * "New cardinality estimation algorithms for HyperLogLog sketches"
+   * Otmar Ertl, arXiv:1702.01284 */
+  double z = m * hllTau((m - reghisto[HLL_Q + 1]) / (double)m);
+  for (j = HLL_Q; j >= 1; --j) {
+    z += reghisto[j];
+    z *= 0.5;
+  }
+  z += m * hllSigma(reghisto[0] / (double)m);
+  E = llroundl(HLL_ALPHA_INF * m * m / z);
+
+  return (uint64_t)E;
+}
+
+/* Merge by computing MAX(registers[i],hll[i]) the HyperLogLog 'hll'
+ * with an array of uint8_t HLL_REGISTERS registers pointed by 'max'.
+ *
+ * The hll object must be already validated via isHLLObjectOrReply()
+ * or in some other way.
+ *
+ * If the HyperLogLog is sparse and is found to be invalid, C_ERR
+ * is returned, otherwise the function always succeeds. */
+void Hyperloglog::hllMerge(uint8_t *max, const std::vector<uint8_t> &counts) {
+  int i;
+  uint8_t val;
+  auto registers = (uint8_t *)(&counts[0]);
+
+  for (i = 0; i < HLL_REGISTERS; i++) {
+    HLL_DENSE_GET_REGISTER(val, registers, i);
+    if (val > max[i]) max[i] = val;
+  }
+}
+
+rocksdb::Status Hyperloglog::getRegisters(const Slice &user_key, std::vector<uint8_t> *counts) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  HyperloglogMetadata metadata;
+  rocksdb::Status s = GetMetadata(ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
+
+  std::string prefix = InternalKey(ns_key, "", metadata.version, storage_->IsSlotIdEncoded()).Encode();
+  std::string next_version_prefix = InternalKey(ns_key, "", metadata.version + 1, storage_->IsSlotIdEncoded()).Encode();
+
+  rocksdb::ReadOptions read_options;
+  LatestSnapShot ss(storage_);
+  read_options.snapshot = ss.GetSnapShot();
+  rocksdb::Slice upper_bound(next_version_prefix);
+  read_options.iterate_upper_bound = &upper_bound;
+
+  auto iter = util::UniqueIterator(storage_, read_options);
+  for (iter->Seek(prefix); iter->Valid() && iter->key().starts_with(prefix); iter->Next()) {
```

Comment:
Can consider using `MultiGet` while the key's indexes are known, guess the performance will be better than the seek way.